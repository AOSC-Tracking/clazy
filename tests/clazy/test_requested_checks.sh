unset CLAZY_CHECKS
unset CLAZY_FIXIT
CLAZY_COMMAND="clazy -c -o /dev/null -xc++ -Xclang -plugin-arg-clang-lazy -Xclang print-requested-checks "
CLAZY_COMMAND_STDIN=$CLAZY_COMMAND"-"

# Test without checks:
echo | $CLAZY_COMMAND_STDIN


# Test with invalid check:
export CLAZY_CHECKS="foo"
echo | $CLAZY_COMMAND_STDIN

# Test with 1 check specified through env variable
export CLAZY_CHECKS="foreach"
echo | $CLAZY_COMMAND_STDIN

# Test with 2 checks specified through env variable
export CLAZY_CHECKS="foreach,writing-to-temporary"
echo | $CLAZY_COMMAND_STDIN

# Test with 2 checks specified through env variable plus one error
export CLAZY_CHECKS="foreach,writing-to-temporary,foo"
echo | $CLAZY_COMMAND_STDIN

# Test that fixit enables the check
unset CLAZY_CHECKS
export CLAZY_FIXIT="fix-old-style-connects"
echo | $CLAZY_COMMAND_STDIN

# Test both check and fixit
export CLAZY_CHECKS="old-style-connect"
export CLAZY_FIXIT="fix-old-style-connects"
echo | $CLAZY_COMMAND_STDIN

# Test fixit+check + unrelated check
export CLAZY_CHECKS="old-style-connect,foreach"
export CLAZY_FIXIT="fix-old-style-connects"
echo | $CLAZY_COMMAND_STDIN

# test all_checks
unset CLAZY_FIXIT
export CLAZY_CHECKS="all_checks"
echo | $CLAZY_COMMAND_STDIN

unset CLAZY_FIXIT
unset CLAZY_CHECKS

# Test specifying check in command line
echo | $($CLAZY_COMMAND" -Xclang -plugin-arg-clang-lazy -Xclang implicit-casts -")

# Pass two checks in command line
echo | $($CLAZY_COMMAND" -Xclang -plugin-arg-clang-lazy -Xclang implicit-casts,foreach -")