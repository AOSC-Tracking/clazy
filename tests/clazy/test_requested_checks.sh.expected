Requested checks: auto-unexpected-qstringbuilder, child-event-qobject-cast, connect-non-signal, container-anti-pattern, copyable-polymorphic, detaching-temporary, foreach, inefficient-qlist-soft, lambda-in-connect, missing-qobject-macro, mutable-container-key, non-pod-global-static, post-event, qdatetime-utc, qdeleteall, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-left, qstring-ref, qt-macros, qvariant-template-instantiation, range-loop, returning-data-from-temporary, rule-of-two-soft, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Invalid check: foo
Requested checks: auto-unexpected-qstringbuilder, child-event-qobject-cast, connect-non-signal, container-anti-pattern, copyable-polymorphic, detaching-temporary, foreach, inefficient-qlist-soft, lambda-in-connect, missing-qobject-macro, mutable-container-key, non-pod-global-static, post-event, qdatetime-utc, qdeleteall, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-left, qstring-ref, qt-macros, qvariant-template-instantiation, range-loop, returning-data-from-temporary, rule-of-two-soft, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: foreach
Requested checks: foreach, writing-to-temporary
Invalid check: foo
Requested checks: foreach, writing-to-temporary
Requested checks: old-style-connect
Requested checks: old-style-connect
Requested checks: foreach, old-style-connect
Requested checks: auto-unexpected-qstringbuilder, base-class-event, child-event-qobject-cast, connect-non-signal, container-anti-pattern, container-inside-loop, copyable-polymorphic, ctor-missing-parent-argument, detaching-temporary, foreach, function-args-by-ref, function-args-by-value, global-const-char-pointer, implicit-casts, incorrect-emit, inefficient-qlist-soft, lambda-in-connect, missing-qobject-macro, missing-typeinfo, mutable-container-key, non-pod-global-static, old-style-connect, post-event, qdatetime-utc, qdeleteall, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-allocations, qstring-arg, qstring-insensitive-allocation, qstring-left, qstring-ref, qt-macros, qvariant-template-instantiation, range-loop, reserve-candidates, returning-data-from-temporary, returning-void-expression, rule-of-three, rule-of-two-soft, temporary-iterator, unused-non-trivial-variable, virtual-call-ctor, writing-to-temporary, wrong-qglobalstatic
Requested checks: implicit-casts
Requested checks: foreach, implicit-casts
Requested checks: old-style-connect
Requested checks: connect-non-signal, container-anti-pattern, lambda-in-connect, mutable-container-key, qdatetime-utc, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-ref, qt-macros, qvariant-template-instantiation, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: auto-unexpected-qstringbuilder, child-event-qobject-cast, connect-non-signal, container-anti-pattern, copyable-polymorphic, detaching-temporary, foreach, inefficient-qlist-soft, lambda-in-connect, missing-qobject-macro, mutable-container-key, non-pod-global-static, post-event, qdatetime-utc, qdeleteall, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-left, qstring-ref, qt-macros, qvariant-template-instantiation, range-loop, returning-data-from-temporary, rule-of-two-soft, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: connect-non-signal, container-anti-pattern, lambda-in-connect, mutable-container-key, qdatetime-utc, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-ref, qt-macros, qvariant-template-instantiation, reserve-candidates, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: connect-non-signal, container-anti-pattern, lambda-in-connect, mutable-container-key, qdatetime-utc, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-ref, qt-macros, qvariant-template-instantiation, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: connect-non-signal, container-anti-pattern, foreach, implicit-casts, lambda-in-connect, mutable-container-key, qdatetime-utc, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-ref, qt-macros, qvariant-template-instantiation, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: auto-unexpected-qstringbuilder, child-event-qobject-cast, connect-non-signal, container-anti-pattern, copyable-polymorphic, detaching-temporary, foreach, inefficient-qlist-soft, lambda-in-connect, missing-qobject-macro, mutable-container-key, non-pod-global-static, post-event, qdatetime-utc, qdeleteall, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-left, qstring-ref, qt-macros, qvariant-template-instantiation, range-loop, returning-data-from-temporary, rule-of-two-soft, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: connect-non-signal, container-anti-pattern, lambda-in-connect, mutable-container-key, qdatetime-utc, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-ref, qt-macros, qvariant-template-instantiation, reserve-candidates, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: connect-non-signal, container-anti-pattern, implicit-casts, lambda-in-connect, mutable-container-key, qdatetime-utc, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-ref, qt-macros, qvariant-template-instantiation, reserve-candidates, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: implicit-casts
Requested checks: implicit-casts
Could not find checks in comma separated string implicit-casts,no-implicit-casts
Available checks and FixIts:

Checks from level0:
    connect-non-signal
    container-anti-pattern
    lambda-in-connect
    mutable-container-key
    qdatetime-utc    (fix-qdatetime-utc)
    qenums
    qfileinfo-exists
    qgetenv    (fix-qgetenv)
    qmap-with-pointer-key
    qstring-arg
    qstring-insensitive-allocation
    qstring-ref    (fix-missing-qstringref)
    qt-macros
    qvariant-template-instantiation
    temporary-iterator
    unused-non-trivial-variable
    writing-to-temporary
    wrong-qglobalstatic

Checks from level1:
    auto-unexpected-qstringbuilder    (fix-auto-unexpected-qstringbuilder)
    child-event-qobject-cast
    copyable-polymorphic
    detaching-temporary
    foreach
    inefficient-qlist-soft
    missing-qobject-macro
    non-pod-global-static
    post-event
    qdeleteall
    qstring-left
    range-loop
    returning-data-from-temporary
    rule-of-two-soft

Checks from level2:
    base-class-event
    container-inside-loop
    ctor-missing-parent-argument
    function-args-by-ref
    function-args-by-value
    global-const-char-pointer
    implicit-casts
    incorrect-emit
    missing-typeinfo
    old-style-connect    (fix-old-style-connect)
    qstring-allocations    (fix-qlatin1string-allocations,fix-fromLatin1_fromUtf8-allocations,fix-fromCharPtrAllocations)
    reserve-candidates
    returning-void-expression
    rule-of-three
    virtual-call-ctor

Checks from level3:
    assert-with-side-effects
    bogus-dynamic-cast
    detaching-member

If nothing is specified, all checks from level0 and level1 will be run.

To specify which checks to enable set the CLAZY_CHECKS env variable, for example:
    export CLAZY_CHECKS="level0"
    export CLAZY_CHECKS="level0,reserve-candidates,qstring-allocations"
    export CLAZY_CHECKS="reserve-candidates"

or pass as compiler arguments, for example:
    -Xclang -plugin-arg-clang-lazy -Xclang reserve-candidates,qstring-allocations

To enable FixIts for a check, also set the env variable CLAZY_FIXIT, for example:
    export CLAZY_FIXIT="fix-qlatin1string-allocations"

FixIts are experimental and rewrite your code therefore only one FixIt is allowed per build.
Specifying a list of different FixIts is not supported.
Backup your code before running them.
Requested checks: connect-non-signal, container-anti-pattern, lambda-in-connect, mutable-container-key, qdatetime-utc, qfileinfo-exists, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-ref, qt-macros, qvariant-template-instantiation, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: implicit-casts
Requested checks: implicit-casts
Requested checks: auto-unexpected-qstringbuilder, child-event-qobject-cast, connect-non-signal, container-anti-pattern, copyable-polymorphic, detaching-temporary, foreach, inefficient-qlist-soft, lambda-in-connect, missing-qobject-macro, mutable-container-key, non-pod-global-static, post-event, qdatetime-utc, qdeleteall, qenums, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-left, qstring-ref, qt-macros, qvariant-template-instantiation, range-loop, returning-data-from-temporary, rule-of-two-soft, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: connect-non-signal, container-anti-pattern, lambda-in-connect, mutable-container-key, qdatetime-utc, qfileinfo-exists, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-ref, qt-macros, qvariant-template-instantiation, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
Requested checks: auto-unexpected-qstringbuilder, child-event-qobject-cast, connect-non-signal, container-anti-pattern, copyable-polymorphic, detaching-temporary, foreach, inefficient-qlist-soft, lambda-in-connect, missing-qobject-macro, mutable-container-key, non-pod-global-static, post-event, qdatetime-utc, qdeleteall, qfileinfo-exists, qgetenv, qmap-with-pointer-key, qstring-arg, qstring-insensitive-allocation, qstring-left, qstring-ref, qt-macros, qvariant-template-instantiation, range-loop, returning-data-from-temporary, rule-of-two-soft, temporary-iterator, unused-non-trivial-variable, writing-to-temporary, wrong-qglobalstatic
